### PR TITLE
Feature/bng index

### PR DESCRIPTION
**Pull request by milos-colic**
Adding support for BNG.
Some additional minor fixes.

### CODE REVIEW

File: src/main/scala/com/databricks/labs/mosaic/core/index/BNGIndexSystem.scala [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+import scala.util.{Success, Try}
+
+import com.databricks.labs.mosaic.core.geometry.MosaicGeometry
+import com.databricks.labs.mosaic.core.geometry.api.GeometryAPI
+import com.databricks.labs.mosaic.core.types.model.GeometryTypeEnum.POLYGON
+import com.databricks.labs.mosaic.core.types.model.MosaicChip
+import org.locationtech.jts.geom.Geometry
+
+/**
+  * Implements the [[IndexSystem]] via BNG java implementation.
```
Comment (milos-colic) [inline src/main/scala/com/databricks/labs/mosaic/core/index/BNGIndexSystem.scala:13]: @milos-colic more docs needed.

File: src/test/scala/com/databricks/labs/mosaic/expressions/geometry/TestIntersectionExpressionsCodegen.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        it should behave like intersectsCodegen(MosaicContext.build(BNGIndexSystem, ESRI), spark)
+        it should behave like intersectsCodegen(MosaicContext.build(BNGIndexSystem, JTS), spark)
+        it should behave like intersects(MosaicContext.build(H3IndexSystem, ESRI), spark, 11)
+        it should behave like intersects(MosaicContext.build(H3IndexSystem, JTS), spark, 11)
+        it should behave like intersects(MosaicContext.build(BNGIndexSystem, ESRI), spark, 5)
+        it should behave like intersects(MosaicContext.build(BNGIndexSystem, JTS), spark, 5)
+    }
+
+    "ST_IntersectionAggregate" should "compute the intersection via aggregate expression" in {
+        //it should behave like intersectionCodegen(MosaicContext.build(H3IndexSystem, ESRI), spark)
```
Comment (sllynn) [inline src/test/scala/com/databricks/labs/mosaic/expressions/geometry/TestIntersectionExpressionsCodegen.scala:23]: Should these be uncommented?

**Review by sllynn — COMMENTED** [review]
A few Qs in there. Small things I think.

Rather than me pulling this branch and updating the docs, why don't you merge this to main and I'll then pull main into my docs PR and add the new functions there.

File: src/main/scala/com/databricks/labs/mosaic/core/types/ChipType.scala [hunk 3]
```diff
@@ -14,5 +14,9 @@ class ChipType()
         StructField("wkb", BinaryType)
       )
     ) {
+
+    override def simpleString: String = "CHIP"
```
[the remark this answers is not in the document]
Comment (milos-colic) [reply src/main/scala/com/databricks/labs/mosaic/core/types/ChipType.scala:18]: I have tested the python bindings in a notebook.
.show() is not imperative to support .display() is preferred and that worked fine.

File: src/main/scala/com/databricks/labs/mosaic/core/geometry/point/MosaicPointESRI.scala [hunk 4]
```diff
@@ -66,6 +65,18 @@ object MosaicPointESRI extends GeometryReader {
...
     }
 
+    def apply(coords: Seq[Double]): MosaicPointESRI = {
+        if (coords.length == 3) {
+            MosaicPointESRI(
+              new OGCPoint(new Point(coords(0), coords(1), coords(2)), SpatialReference.create(4326))
+            )
+        } else {
+            MosaicPointESRI(
+              new OGCPoint(new Point(coords(0), coords(1)), SpatialReference.create(4326))
```
[the remark this answers is not in the document]
Comment (milos-colic) [reply src/main/scala/com/databricks/labs/mosaic/core/geometry/point/MosaicPointESRI.scala:75]: yes, that is better.

File: src/test/scala/com/databricks/labs/mosaic/expressions/geometry/TestIntersectionExpressionsCodegen.scala (hunk 2 above)
Comment (milos-colic) [reply src/test/scala/com/databricks/labs/mosaic/expressions/geometry/TestIntersectionExpressionsCodegen.scala:23]: yes, they should

**Comment by milos-colic** [comment]
> A few Qs in there. Small things I think.
> 
> Rather than me pulling this branch and updating the docs, why don't you merge this to main and I'll then pull main into my docs PR and add the new functions there.

@sllynn I have fixed the comments in the last commit.
I agree that we should merge this then open another one for docs update since you own the docs side.

File: src/main/scala/com/databricks/labs/mosaic/core/index/BNGIndexSystem.scala [hunk 5]
```diff
@@ -368,16 +476,22 @@ object BNGIndexSystem extends IndexSystem with Serializable {
       */
     override def getResolution(res: Any): Int = {
```
Comment (sllynn) [inline src/main/scala/com/databricks/labs/mosaic/core/index/BNGIndexSystem.scala:477]: As per the slack discussion, need to make sure that all expressions that compute indexes use the getResolution method for the correct index system (PointIndexGeom is still using H3).

Comment (milos-colic) [reply src/main/scala/com/databricks/labs/mosaic/core/index/BNGIndexSystem.scala:477]: fixed in the last commit.

**Review by sllynn — APPROVED** [review]
LGTM. If the tests pass, ship it.

**Pull request merged**